plumber/ppl/lib/ppl/retention/record_deleter_queries.ex

```diff
@@ -0,0 +1,50 @@
+defmodule Ppl.Retention.RecordDeleterQueries do
+  @moduledoc """
+  Database queries for the retention record deleter.
+  """
+
+  import Ecto.Query
+
+  alias Ppl.EctoRepo
+  alias Ppl.PplRequests.Model.PplRequests
+
+  @doc """
+  Deletes up to `limit` expired records (where expires_at < now).
+  First deletes associated blocks, then deletes the ppl_requests.
+  Returns the count of deleted records.
+  """
+  @spec delete_expired_batch(non_neg_integer()) :: {:ok, non_neg_integer()} | {:error, term()}
+  def delete_expired_batch(limit) do
+    now = NaiveDateTime.utc_now()
+
+    expired_records =
+      from(pr in PplRequests,
+        where: pr.expires_at < ^now,
+        select: %{id: pr.id, ppl_artefact_id: pr.ppl_artefact_id},
+        limit: ^limit
+      )
+      |> EctoRepo.all()
```

Comment:
Just a comment, but try to run this with the explain to validate that the planner picks the index that you created to run this query. (It should infer expires_at is not null but it may not)